Formalize most public API

PepperCode1 opened the pull request:
This pull request performs many, mostly minor, changes across most public API to make it nicer to use and more stable in general.

PepperCode1 committed:
Start on general API formalization

Jozufozu commented on fabric/src/lib/java/dev/engine_room/flywheel/lib/model/baked/BakedModelBufferer.java:
```diff
@@ -31,8 +32,18 @@ final class BakedModelBufferer {
-	public static void bufferSingle(ModelBlockRenderer blockRenderer, BlockAndTintGetter level, BakedModel model, BlockState state, @Nullable PoseStack poseStack, ResultConsumer resultConsumer) {
+	public static void bufferSingle(ModelBlockRenderer blockRenderer, @Nullable BlockAndTintGetter level, BakedModel model, @Nullable BlockState state, @Nullable PoseStack poseStack, ResultConsumer resultConsumer) {
 		ThreadLocalObjects objects = THREAD_LOCAL_OBJECTS.get();
+		if (level == null) {
+			if (state == null) {
+				state = Blocks.AIR.defaultBlockState();
+			}
+			FabricOriginBlockAndTintGetter originLevel = objects.level;
+			originLevel.originBlockState(state);
+			level = originLevel;
+		} else if (state == null) {
+			state = level.getBlockState(BlockPos.ZERO);
+		}
```
This seems like complex behavior to hide in the impl. Somewhat related, perhaps there should be explicit methods for defaultLevel and defaultBlockState in the common builder

PepperCode1 replied:
All builder methods for all three types of builders are optional and thus have default values. Would it be better to add `default*()` methods for all properties or accept nullable parameters, where null means default?

Jozufozu commented on common/src/lib/java/dev/engine_room/flywheel/lib/model/baked/PartialModel.java:
```diff
@@ -1,53 +1,49 @@
-		synchronized (ALL) {
-			ALL.add(this);
+		if (populateOnInit) {
+			FlwLibXplat.INSTANCE.getBakedModel(Minecraft.getInstance().getModelManager(), modelLocation);
```
Looks like bakedModel should be assigned here, and probably throw an error if it's missing

PepperCode1 replied:
Good catch. The method returns the missing model if there were errors during loading or the model doesn't exist. In the former case, vanilla already logs a warning/error, so an additional error isn't necessary.

Jozufozu replied:
Ah a missing model would be better yeah

Jozufozu reviewed (changes requested):
Lots of good stuff in here, just a few things that caught my eye

PepperCode1 committed:
More API improvements

- Add Engine#onLightUpdate; remove LightUpdateHolder and backend/ClientChunkCacheMixin
- Add Effect#level
- Add VisualizationHelper#queueAdd and #queueRemove for Effects
- Fix PartialModel not assigning bakedModel field when populating on init
- Fix PartialModel.ALL using weak keys instead of weak values
- Make Simple*Visualizer and corresponding inner Builder classes final
- Restore FlatLit#light overload that accepts block and sky light values separately
- Add AbstractBlockEntityVisual#relight overloads that accept Iterator and Iterable
- Reorganize classes in impl.vizualization

PepperCode1 committed:
TaskExecutor simplification

- Move TaskExecutor#sync* methods to TaskExecutorImpl
- Move Flag and RaisePlan to impl
- Remove TaskExecutor#scheduleForMainThread and #isMainThread methods
- Remove SyncedPlan
- Add Engine#setupRender
- Remove TaskExecutor parameters from Engine#render* methods
- Convert Engine$CrumblingBlock into an interface
- Unmark RenderContext as NonExtendable to allow fulfilling the purpose described in the doc of VisualizationManager#renderDispatcher

Jozufozu reviewed (approved):
Very nice changes ❤️

PepperCode1 committed:
Remove registry freeze callbacks

- Lazily initialize MaterialShaderIndices
- Rename MaterialShaders#*Shader to #*Source
- Move BackendImplemented to api.backend package

PepperCode1 commented:
I will merge the PR now as there are a lot of changes and waiting for more minor changes is not worth the merge conflicts with other mainline changes that this will cause.
Things to review later:
- Complex level and state behavior in model builders
- Default methods for model builders
- Adding or removing data from `RenderContext`; passing a reduced `RenderContext` to `Engine#updateRenderOrigin` with more than just the `Camera`
- More registry refactors, including renames, removing `registerAndGet`, splitting `MaterialShaders`, making shader source locations easier to register, and refactoring and freezing `VisualizerRegistry`
